Email template

mjgui opened the pull request:
## Problem

Current emails are mostly in plaintext, making them seem like phishing emails. Additionally, they carry no branding of the agency that sent the emails.

Closes [#1203, #1216]

## Solution

_Note: HTML theme is still TBC, waiting for design to be finalised_
- Implemented a HTML template for future emails which includes agency name (in the footer) and agency logo
- Updated `agencies` table to include `name` and `logo_uri` columns, add `domains` table to store mappings from domains to agencies
- Added backend logic to lookup agency details based on reply_to value
- Updated existing email preview to use same styles 

## Before & After Screenshots
_Note: Copy in screenshots is outdated_

**Desktop View**

<img src="https://user-images.githubusercontent.com/9080115/121170595-016d9b80-c888-11eb-9559-3afb78489104.png" width="500"/>

**Mobile View**
<img src="https://user-images.githubusercontent.com/9080115/121160756-f19d8980-c87e-11eb-8a9d-ff344e1218f5.png" width="300"/>

**Rich text Edtior**
<img src="https://user-images.githubusercontent.com/9080115/122186539-b6bfd500-cec0-11eb-9429-0af23b9e0846.png" width="500"/>

## Tests

- Send out any email - email should be sent using the new theme
- Draft a new email campaign - rich text editor should be using Arial font (in black), email preview (in step 2 and 4) should show the new theme

## Deploy Notes

- Motivation behind new schema is [here](https://github.com/opengovsg/postmangovsg/issues/1203#issuecomment-852788863)
_Note: If SQL migrations PR is deployed beforehand, will modify scripts below to use migrations instead_

**Pre-deployment Checklist**
- [ ] Remove pri key constraint from `domain` column (will be unused), add new columns to `agencies` table
```sql
ALTER TABLE agencies 
  DROP CONSTRAINT agencies_pkey,
  ALTER COLUMN domain DROP NOT NULL,
  ADD COLUMN id serial PRIMARY KEY,
  ADD COLUMN name varchar,
  ADD COLUMN logo_uri text;
```
**Note:** Do not drop `domain` column yet to prevent breaking old code

- [ ] Create new `domains` table
```sql
CREATE TABLE domains (
  domain varchar NOT NULL PRIMARY KEY,
  agency_id integer,
  created_at timestamptz,
  updated_at timestamptz,
  FOREIGN KEY (agency_id) REFERENCES agencies(id)
);
```

- [ ] Add `email_domain` column to User (FK on domain)
```sql
ALTER TABLE users
  ADD COLUMN email_domain varchar,
  ADD FOREIGN KEY (email_domain) REFERENCES domains(domain);
```

**When we want to add Logo support**
Note: For now, we can skip the following two steps (populate agencies and logos) until we finalise the logos. This PR only needs the above schema changes to run and will work even if the new tables are blank.

- [ ] Populate list of agencies and domains, populate existing users with domains

- [ ] After populating domains, backpatch existing users table with domains (if valid)
```sql
UPDATE users 
  SET email_domain = d.domain
  FROM domains d
  WHERE SUBSTR(users.email, POSITION( '@' in users.email ) )  = d.domain;
```

**After Deployment is successful (i.e. no more instances of code depending on the old schema)**
- [ ] Drop `domain` column from `agencies` table
_(Optional: Leave domain column for schema compatibility with previous versions of code)_
```sql
ALTER TABLE agencies DROP domain;
```

**New dev dependencies**:

- Added `copyfiles` package to `/shared` to copy static files (email template) to build folder - manually triggered in Dockerfile durin g build

lamkeewei commented on shared/src/theme/index.ts:
```diff
@@ -0,0 +1,17 @@
+  agencyLogoURI?: string
+  unsubLink: string
+}): string => {
+  const htmlTemplate = fs.readFileSync(path.resolve(__dirname, './email-theme.mustache'), 'utf-8')
```
Might be better to use the async version of `readFile`. Synchronous read blocks the entire Node.js event loop.

mjgui replied:
Makes sense, updated

mjgui commented on shared/src/theme/index.ts:
```diff
@@ -0,0 +1,20 @@
+  agencyLogoURI?: string
+  unsubLink: string
+}): Promise<string> => {
+  const htmlTemplate = await fs.promises.readFile(
+    path.resolve(__dirname, './email-theme.mustache'),
+    'utf-8'
+  )
```
Actually, should I move this out as a global constant? Seems wiser to only read the file once during initial run instead of for every message.
Or alternatively, refactor this entire file as a singleton class so it will get loaded a max of 1 time but will not be loaded if it isn't used.

lamkeewei replied:
Makes sense. Though slight preference for the second method. Ideally, importing a module should not result in any side effect.

lamkeewei commented on backend/Dockerfile:
```diff
@@ -13,7 +13,7 @@ RUN aws configure set default.region ap-southeast-1
 
 COPY shared ../shared
-RUN cd ../shared && npm ci && cd ../postmangovsg
+RUN cd ../shared && npm ci
```
```suggestion
RUN cd ../shared && npm ci && npm run build
```

This should work too (haven't tested). Originally I didn't have to run `npm run build` in `shared` because the tsc command in `backend` will build all project references also. But in this case since there's post-build actions might be easier to explicitly run build.

mjgui replied:
Yep I was thinking about this, just that this will result in `shared` being built twice. I doubt it will have significant cost, but not sure if calling `npm run build` (and hence implicitly calling the postbuild script) is more readable or if running `postbuild` is more readable

lamkeewei commented on backend/src/email/services/email.service.ts:
```diff
@@ -61,11 +65,33 @@ const getHydratedMessage = async (
+      },
+    ],
+  })
+  const agency = campaign?.user?.domain?.agency
```
Is there a situation where agency name or agency logo will be `undefined` and the hydrated message will still be considered valid? If not, it would be good to do a check here and throw an error if they are undefined. 

This way you can remove the `undefined` type for `agencyName` and `agencyLogoURI`. Benefits are that callers of this method no longer need to check for `undefined` from the return value and it better encodes the domain.

mjgui replied:
Actually yes, the database will allow for `NULL` values (which I guess will become `undefined` through sequelize?).

For `agencyLogoURI` for example, we're currently planning to leave it blank for agencies who don't have suitable logos (and potentially for the generic government agency too)

For `agencyName`, a useful side effect for leaving it optional as well is that if either `domain` or `agency` is undefined (such as for users in the current DB before backpatching), the new code will run perfectly too

mjgui commented on shared/src/theme/index.ts:
```diff
@@ -0,0 +1,60 @@
+        whiteList: {},                  // whitelist for tags is empty, i.e. strip all HTML tags 
+        stripIgnoreTagBody: ['script'], // remove contents of script tags entirely
+        onIgnoreTag: (_tag, _html, options) => {
+          // @ts-ignore types are missing, upgrade to js-xss 1.0.9 which fixes missing types
```
had to leave a @ts-ignore here because the types for `js-xss` library are incomplete, see [this issue](https://github.com/leizongmin/js-xss/issues/224)

lamkeewei commented:
@mjgui sequelize migrations have been merged in. Would be good to change to use migrations instead of the manual SQL.

lamkeewei commented:
Probably want to use a different color for the placeholder text.

![image](https://user-images.githubusercontent.com/3666479/122345649-a9672100-cf7a-11eb-9c38-bca0187a3bbb.png)

mjgui commented:
Removed all DB schema changes (and rebased on the latest `develop` branch) at #1242 
Will re-add schema changes after confirmation on implementation of logos
